### ISSUE TITLE
Automatic detection of `choices` in `match.arg()`

### ISSUE THREAD

**Issue by Bisaloo** [epinowcast/epinowcast#156]
The base function `match.arg()` is used throughout the code to check that values provided for arguments are valid input. Each time, the `choice` argument to indicate what values are valid. Some examples:

https://github.com/epiforecasts/epinowcast/blob/603782bece7cea70d68287d4481fc0dc08a006b8/R/formula-tools.R#L250

https://github.com/epiforecasts/epinowcast/blob/603782bece7cea70d68287d4481fc0dc08a006b8/R/methods.R#L23-L26

This is not strictly necessary since R already includes a mechanism to automatically detect the value of this choice argument:

``` r
f <- function(arg1 = c("value1", "value2")) {
  
  # This line is equivalent to:
  # arg1 <- match.arg(arg1, choices = c("value1", "value2"))
  arg1 <- match.arg(arg1)
  
  return(arg1)
  
}

f()
#> [1] "value1"
f("value1")
#> [1] "value1"
f("value2")
#> [1] "value2"
f("wronginput")
#> Error in match.arg(arg1): 'arg' should be one of "value1", "value2"
```

<sup>Created on 2022-08-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Should `choices` be provided in `match.arg()` calls in epinowcast or should it rely on the automatic detection feature?**

Some pros/cons discussed with @seabbs this morning:

- Current style:
  * the list of choices is visible on the line where `match.arg()` is called rather than possibly having to scroll up to the function definition
  * slightly faster
 
- Proposed style:
  * the list of choices is visible in the documentation (although it is good practice to re-document it in the relevant roxygen2 `@param` section)
  * slightly more idiomatic. This is used throughout base R code as well as commonly used packages such as the tidyverse.

**Comment by Bisaloo** [comment]
To weigh in, I don't think the two arguments in favour of the current style are very strong:

> the list of choices is visible on the line where match.arg() is called rather than possibly having to scroll up to the function definition

`match.arg()`, as every other input checking, should almost always be called right at the start of the function.

> slightly faster

It is faster in benchmarks but it is not likely to be significant compared to the rest of the function. This step will always be fast, no matter what style is chosen.


Overall, I would lean towards the more idiomatic option.

**Comment by seabbs** [comment]
Thanks again Hugo and thanks for making such a clear issue (perhaps we should link to this from the contributing guide as a good example). 

Not really being aware of the trade-offs before and with some thought I also agree we should switch to your suggested appoach. I also don't think the arguments in favour are that strong. 

Out of interest do you know why this works? It almost seems like magic? Something about capturing the default args in the environment when the function is created?

**Comment by pearsonca** [comment]
can see how it works with `> match.args` - all the bits implemented in R, no compiled call outs. happens right at the top.

uses `formals` to get at function as-defined, then does some introspection.

agree w/ @Bisaloo re putting the options in function definition.

**Comment by seabbs** [comment]
Closed in #162
